reporter.go

```diff
@@ -0,0 +1,302 @@
+	type processedTestCase struct {
+		TestCase jUnitTestReportSchema
+		Error    error
+	}
+
+	testChan := make(chan processedTestCase, len(*tests))
+
+	// Limit the maximum concurrent requests
+	httpThreads := make(chan struct{}, reporterConcurrency)
+
+	processTest := func(test rainforest.RunTestDetails) {
+		testCase := jUnitTestReportSchema{}
+
+		duration, err := timeStringDifferenceSecs(test.CreatedAt, test.UpdatedAt)
+
+		if err != nil {
+			testChan <- processedTestCase{TestCase: testCase, Error: err}
+			return
+		}
+
+		testCase.Name = test.Title
+		testCase.Time = duration
+
+		if test.Result == "failed" {
+			var testDetails *rainforest.RunTestDetails
+
+			// reserve a thread
+			httpThreads <- struct{}{}
+			log.Printf("Fetching information for failed test #" + strconv.Itoa(test.ID))
+			testDetails, err = api.GetRunTestDetails(runID, test.ID)
+			// release the thread
+			<-httpThreads
+
+			if err != nil {
+				testChan <- processedTestCase{TestCase: jUnitTestReportSchema{}, Error: err}
+				return
+			}
+
+			for _, step := range testDetails.Steps {
+				for _, browser := range step.Browsers {
+					browserName := browser.Name
+
+					for _, feedback := range browser.Feedback {
+						if feedback.AnswerGiven == "no" && feedback.JobState == "approved" && feedback.Note != "" {
+							reportFailure := jUnitTestReportFailure{Type: browserName, Message: feedback.Note}
+							testCase.Failures = append(testCase.Failures, reportFailure)
+						}
+					}
+				}
+			}
+		}
+
+		testChan <- processedTestCase{TestCase: testCase}
+	}
+
+	for _, test := range *tests {
+		go processTest(test)
+	}
+
+	testCases := []jUnitTestReportSchema{}
+
+	for i := 0; i < len(*tests); i++ {
+		processed := <-testChan
+
+		if processed.Error != nil {
+			return &[]jUnitTestReportSchema{}, processed.Error
+		}
+
+		testCases = append(testCases, processed.TestCase)
+	}
+
+	return &testCases, nil
+}
+
+func writeJUnitReport(reportSchema *jUnitReportSchema, file *os.File) error {
+	enc := xml.NewEncoder(file)
+
+	file.Write([]byte(xml.Header))
+
+	enc.Indent("", "  ")
+	err := enc.Encode(reportSchema)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("JUnit report successfully written to %v", file.Name())
+
+	file.Close()
+	return nil
+}
+
+/*
+	Utility Functions
+*/
+
+func createOutputFile(filepath string) (*os.File, error) {
+	var file *os.File
+	var err error
+	if file, err = os.Create(filepath); err != nil {
+		return file, err
+	}
+	return file, err
+}
+
+func timeStringDifferenceSecs(start string, end string) (float64, error) {
+	var startTime time.Time
+	var endTime time.Time
+	var err error
+	var diff float64
+
+	if startTime, err = time.Parse(time.RFC3339Nano, start); err != nil {
+		return diff, err
+	}
+
+	if endTime, err = time.Parse(time.RFC3339Nano, end); err != nil {
+		return diff, err
+	}
+
+	diff = endTime.Sub(startTime).Seconds()
+
+	return diff, nil
+}
```

Comment:
If you change the types to be timestamps in a api struct, this function may also prove to be unnecessary.